src/rlx_assemble.erl

```diff
@@ -1086,5 +1092,13 @@ format_error({strip_release, Reason}) ->
 format_error({rewrite_app_file, AppFile, Error}) ->
     io_lib:format("Unable to rewrite .app file ~s due to ~p",
                   [AppFile, Error]);
+format_error({consult_app_file, AppFile, enoent}) ->
+    io_lib:format("Unable to read .app file ~s (file not found).~n"
+                  "If the app and hex package are named differently, you will need to specify the~n"
+                  "dependency in rebar.config with {AppName, {pkg, PackageName}}",
```

Comment:
I don't think this is the right way to go. There should already be a failure well before we get to this point. Something is wrong earlier if rebar3 is somehow able to compile the project even though the name of a dep package and its application differ.

Reply:
I don't know that it checks it. Like it could internally just go "yeah this is one of my apps" and put it in path, and you would likely get a failure only once you hit a task that checks and fetches specific apps by name: releases, dialyzer, escriptize. 

The compiler just goes "what the hell this is all correct" and so does the locking mechanism, happily locking the wrong app name as declared by the deps. The cross-checking is done once you try to load an app by the dep name rather than by the actual app name if there's any mismatch.

Reply:
Hm, so yes but also, not the same fix in that case. You are saying in a case htat you put `PackageName` in the `.app.src` file list of applications? Then yea, that probably only fails when it gets to trying to load the runtime deps like in `rebar3 shell` or `release`.

But in that case the issue isn't that the user needs to put `{AppName, {pkg, PackageName}}` in `rebar.config`.

Reply:
In the case I've mentioned, there was no adding of the PackageName in the applications list in `.app.src`, however, specifying `{AppName, {pkg, PackageName}}` does indeed fix it.

And if that's not the proper fix, what is the reason for having that option at all?  In the [docs](https://rebar3.org/docs/configuration/dependencies/#declaring-dependencies), the specified purpose of `{AppName, {pkg, PkgName}}` is precisely for apps that are named differently from their packages.

Reply:
That is the purpose, I'm just saying I think it should be being caught earlier and am actually not sure how you got it to get to this point to fail and not fail on `rebar3 compile`.

What did you have in `deps` before you added `{AppName, {pkg, PackageName}}`? You didn't add `PackageName` to `.app.src` so you had `AppName` in `applications` of your apps `.app.src`?

Reply:
Gotcha, Understood. 

`deps` was empty, and I added neither the AppName nor the PackageName to `applications` of `.app.src`.

I did this with a bare project, easily reproducible:

1) `rebar3 new app test_app`
2) edit `test_app/rebar.config`and add `erl_pipeline` to `deps`
3) `rebar3 compile` (succeeds)
4) Set up super basic release with
```
{relx, [{release, {test_app, "0.0.1"},
    [{erl_pipeline, none}, test_app]
}]}.
```
5) `rebar3 release` (crashes)

Reply:
Oh, but wouldn't this crash the same way even if the package name wasn't different from the app name?

Reply:
No, if I revise the deps to be `{pipeline, {pkg, erl_pipeline}}` it works just fine.

Also, I may have just been an idiot, but if `PackageName` (alone) is specified in `deps`, while `AppName` is specified in the `relx` application list, everything works as expected.  Honestly, I thought I had tried that combination and it failed, but I must not have.

That said, I would revise my proposal to remove the recommendation for `{AppName, {pkg, PkgName}}`, however, given the confusing error message, it may be helpful to at least still just show that `filename.app` was not found as the reason for the crash.

Reply:
Revise the deps? You said the deps were empty.

Reply:
No, I don't think I was clear.  I meant with a blank project, with empty deps, adding `{pipeline, {pkg, erl_pipeline}}` worked, while adding just `erl_pipeline` didn't.

I was wrong, in that assertion, however.

The error I'm encountering here is really just if, I as a naive developer, add to the same `PackageName` to both `deps` and `relx`, that will produce the crash.

In order for this to work, I need to know that `PackageName` and `AppName` are not the same, and have `PackageName` in `deps` and `AppName` in `relx`, then everything works.

But the former strikes me as something that can easily be done if the user doesn't realize the app name and the package name aren't the same.  In that case, producing a useful error message would help the user out.

Does that make sense?

Reply:
Yea, the issue definitely makes sense, I just think it is an issue further up the stack.

I could have swore that `{deps, [PackageName]}` resulted in an error during compilation because rebar3 verifies the app name. So the issue should be caught earlier and then we can give a more useful error message because it has more context into what is wrong.

I see this isn't the case, `{deps, [erl_pipeline]}` compiles fine. I think this needs to be where the fix is. I may be wrong that it used to error if this happened, but I still think it is the better place for the fix.

And this PR should only have an error message saying that the appfile doesn't exist, but not anything about package name. The error message when rebar3 catches that the dep fetched fine and built but has a different app name is where it can mention use of `{pkg, erl_pipeline}`.

Reply:
I'm almost positive rebar2 would error out if the app name didn't match the package name, but I'm not sure about rebar3.

Thanks, I'll revise this PR accordingly, removing the reference to the package name and rebar.